### DISCUSSION TITLE
Why fuse_session_loop_mt32 can have two different parameters in definition and declaration?

### DISCUSSION THREAD

**Discussion by yangyun50** [libfuse/libfuse#1015] [Q&A] [answered]
libfuse has declaration of `fuse_session_loop_mt_32` with parameter `struct fuse_loop_config`. However, the definition of `fuse_session_loop_mt_32` has parameter `struct fuse_loop_config_v1`. How can this work? I have written a demo of the same situation, but the compiler report errors:

```C
#include <stdio.h>

struct a_v1;
struct a {
        int d;
};

void add(struct a *x);

void add(struct a_v1 *x)  {
        printf("can compile:%d \n", x->d);
}

int main() {
        struct a y;
        y.d =100;
        add(&y);
        return 0;
}

```

**Comment by bsbernd** [comment]
I guess you mean
```
#if FUSE_USE_VERSION < 32
    int fuse_session_loop_mt_31(struct fuse_session *se, int clone_fd);
    #define fuse_session_loop_mt(se, clone_fd) fuse_session_loop_mt_31(se, clone_fd)
#elif FUSE_USE_VERSION < FUSE_MAKE_VERSION(3, 12)
    int fuse_session_loop_mt_32(struct fuse_session *se, struct fuse_loop_config *config);
    #define fuse_session_loop_mt(se, config) fuse_session_loop_mt_32(se, config)
#else
```
?

Also see declarations in  `include/fuse_common.h`. And also in `lib/fuse_i.h`.
This is a bit tricky, order to use the more recent version of fuse_loop_config for libfuse >= 3.12 (which also treats this struct as private outside of libfuse).

**Reply by yangyun50** [reply]
Thanks for your reply! It seems I didn't describe it clearly.  

Let's assume the libfuse == 3.5. For example, I'm using a passthrough_hp example which defines FUSE_USE_VERSION 35, this means that the function `fuse_session_loop_mt()` will expand to `fuse_session_loop_mt_32(struct fuse_session *se, struct fuse_loop_config *config)` due to the following definition:
```C
#if FUSE_USE_VERSION < 32
    int fuse_session_loop_mt_31(struct fuse_session *se, int clone_fd);
    #define fuse_session_loop_mt(se, clone_fd) fuse_session_loop_mt_31(se, clone_fd)
#elif FUSE_USE_VERSION < FUSE_MAKE_VERSION(3, 12)
    int fuse_session_loop_mt_32(struct fuse_session *se, struct fuse_loop_config *config);
    #define fuse_session_loop_mt(se, config) fuse_session_loop_mt_32(se, config)
```
You can see that the second parameter here is `struct fuse_loop_config`, not `struct fuse_loop_config_v1`，but I didn't find that libfuse has a function definition of `int fuse_session_loop_mt_32(struct fuse_session *se, struct fuse_loop_config *config)` . It only has a function definition of ` int fuse_session_loop_mt_32(struct fuse_session *se, struct fuse_loop_config_v1 *config_v1)`. 

So how does `int fuse_session_loop_mt_32(struct fuse_session *se, struct fuse_loop_config *config)` get converted into `int fuse_session_loop_mt_32(struct fuse_session *se, struct fuse_loop_config_v1 *config_v1)`?  I know there may be a bit tricky. But I don't know how, since I use a similar example as described below, it cannot compile successfully (the function `void add(struct a *x)` cannot get converted into `void add(struct a_v1 *x);`):
```C
#include <stdio.h>

struct a_v1;
struct a {
        int d;
};

void add(struct a_v1 *x);
void add(struct a *x);

void add(struct a_v1 *x)  {
        printf("can compile:%d \n", x->d);
}

int main() {
        struct a y;
        y.d =100;
        add(&y);
        return 0;
}
```
Is this a tricky provided by symbol version?

**Comment by bsbernd** [comment]
passthrough_hp.cc has `#define FUSE_USE_VERSION FUSE_MAKE_VERSION(3, 12)` current master.

In fuse_common.h
```
#if FUSE_USE_VERSION < FUSE_MAKE_VERSION(3, 12)
struct fuse_loop_config_v1; /* forward declarition */
struct fuse_loop_config {
#else
struct fuse_loop_config_v1 {
#endif
```

(`declarition` is obviously a typo)

Your example just doesn't match what the code does - there are two definitions of `struct fuse_loop_config`. One in fuse_common.h for API < 3.12 and the other in fuse_i.h for API >= 3.12.

And then fuse_loop_config_v1 is defined only when API >= 3.12 is used, to be able to convert to the new struct version.

**Reply by yangyun50** [reply]
What if I use an old version (e.g., define FUSE_USE_VERSION = FUSE_MAKE_VERSION(3, 5) in the release 3.10.5) passthrough_hp, and combined with a new version (e.g.,  define FUSE_USE_VERSION = FUSE_MAKE_VERSION(3, 12))  libfuse? My example  is corresponding to this case. I have replaced the passthrough_hp in current libfuse with it  in release 3.10.5 without modification. And of course it woks fine.

But in this case the code goes to `fuse_session_loop_mt()`->declaration of `int fuse_session_loop_mt_32(struct fuse_session *se, struct fuse_loop_config *config);`->definition of `fuse_session_loop_mt_32(struct fuse_session *se, struct fuse_loop_config_v1 *config_v1)`  with the definition of `struct fuse_loop_config`, not `struct fuse_loop_config_v1`. How ?

Besides, libfuse only has declaration of `int fuse_session_loop_mt_32(struct fuse_session *se, struct fuse_loop_config *config)`, but no definition of `int fuse_session_loop_mt_32(struct fuse_session *se, struct fuse_loop_config *config)` . How does it works ?

**Comment by bsbernd** [comment]
There is basically a conversion issue - conversion should have checked from the beginning if a v2 struct was used - in this case conversion could have aborted. As it did not abort, this PR https://github.com/libfuse/libfuse/pull/1018 introduces a warning message and adjusts conversion.

**Comment by bsbernd** [comment, marked as answer]
fuse_session_loop_mt_32() is declared in fuse_loop_mt.c, which gets compiled with the new API/ABI (>=3.12) and `struct fuse_loop_config_v1`.
Applications that link to it get a different declaration for compatibility - `struct fuse_loop_config`. 

And now read in fuse_common.h 
```
#if FUSE_USE_VERSION < FUSE_MAKE_VERSION(3, 12)
struct fuse_loop_config_v1; /* forward declarition */
struct fuse_loop_config {
#else
struct fuse_loop_config_v1 {
#endif
```

I.e. `struct fuse_loop_config` and `struct fuse_loop_config_v1` are binary compatible when the <3.12 API/ABI is used.

**Reply by yangyun50** [reply]
That's what I mean. Got it! Thank you so much. 
I'm sorry for my poor expression and compile knowledges.

**Comment by bsbernd** [comment]
Don't worry, I know it is really tricky and hard to understand. I had spent quite some time to get that working. And thanks for looking at it! If you have some ideas how to make it easier to read - more comments - any suggestion or patch is highly welcome.
